Change top nav from accordion to two rows on mobile

jonahtanjz opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Resolves #980 

Remove current accordion for top navigation on mobile and shift it to create a second row sliding menu. `slot="right"` item will now be visible and positioned on the right side instead of being in an accordion.

Before:
![image](https://user-images.githubusercontent.com/47470981/144993979-e3b8b46a-6d59-4233-95f6-b64179adee17.png)

After:
![image](https://user-images.githubusercontent.com/47470981/144994431-936e0d61-4f74-4ef6-a1b6-2582f0adb642.png)


**Anything you'd like to highlight / discuss:**
NIL

**Testing instructions:**
Preview markbind site with updated top nav design: https://deploy-preview-1690--markbind-master.netlify.app/index.html


<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Change top nav from accordion to two rows on mobile

The default slot and items in the right slot are currently
hidden in an accordion on mobile.

Let's shift the default slot items to a new row so that they are
more accessible and right slot items can now be placed on the
right side of the top nav.

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

ang-zeyu commented:
Nice!

Reviewing the design first; I think it looks good overall, just 2 parts that need to be fixed:
- the horizontal scrollbar dosen't seem to be functional on desktop chrome with mouse scroll wheels (works fine for touchscreens)
-  Could we preserve the look of the dropdown? (absolutely-positioned with respect to the "Dropdown" text so it dosen't add to the height of the scrollbar) (including the searchbar, which uses some common css classes iirc)
   ![Untitled](https://user-images.githubusercontent.com/3306138/145512617-bf0abd9e-376f-404f-ac77-cd46e8be6bbe.png)

jonahtanjz commented:
Hi @ang-zeyu thanks for the review. I've made the changes accordingly.

> the horizontal scrollbar dosen't seem to be functional on desktop chrome with mouse scroll wheels (works fine for touchscreens)

I believe to scroll horizontally with the mouse wheel can hold down `SHIFT` and scroll but this doesn't really seem intuitive 😅 I've bind the mouse wheel to the horizontal scroll on the mobile navbar and this should resolve the issue.

> Could we preserve the look of the dropdown? (absolutely-positioned with respect to the "Dropdown" text so it dosen't add to the height of the scrollbar) (including the searchbar, which uses some common css classes iirc)

Yup agree that it will look better. I've updated the design so that it is now absolute positioned and also added scrolling if the dropdown overflows the bottom of the screen.

![image](https://user-images.githubusercontent.com/47470981/145663895-5a7ed298-92f4-4cb0-863e-4c5f96f7372b.png)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+        }
+
+        .navbar-right {
+            position: absolute;
```
try the flex [`order`](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Ordering_Flex_Items) property; with the right combination you should be able to remove the reliance on `absolute` positioning (not too good in this case as its sizing should be taken into account in the document flow, e.g. if `navbar-brand` is empty this overflows the container)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+            position: absolute;
+            top: 14px;
+            right: 16px;
+            max-width: 40vw;
```
`.navbar-brand` might need this as well.

could we use `50%` too? ( so each part takes up at most half then wraps vertically if needed )

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+        }
+
+        .navbar-default > ul > li,
+        .navbar-default > ul > div {
```
how about just `.navbar-default > ul > *`? since there's no restriction on what can be placed in there 😃

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+        .navbar-default > ul > div {
+            padding: 0.3125rem 0.625rem;
+            flex: 1;
+            background: rgba(0, 0, 0, 0.3);
```
could be moved to `.navbar-default`, so items can set their own backgrounds

jonahtanjz replied:
Hmmm I was thinking that if a link on the navbar is highlighted, its background should also change to the main colour of the navbar. If we move this to `.navbar-default`, then I don't think we can do the background highlighting for the individual link. Any suggestion?

![image](https://user-images.githubusercontent.com/47470981/145839140-e4398162-5d57-4ff3-9d49-5a96e0354609.png)

jonahtanjz replied:
> The alternative I had in mind was to move the said color to .navbar-default, then forward the navbar main colour to .current specifically. This means forwarding .navbar-light/dark/primary's colours to the .current element manually.

I've now shifted the background overlay to the default navbar instead of the individual link. I've also forwarded the navbar colour to a css variable for `.current` to use.

> One slight issue with 0.3 specifically though - you'll likely need to try out / set a different value for .navbar-light, 0.3 seems a little dark for it.

Yup agree with this. I've decreased the overlay for primary and dark to `0.2` and `0.1` for light. Should be better now.

Primary (`0.2`):
![image](https://user-images.githubusercontent.com/47470981/146646323-4be40dae-7fbc-4d82-abea-b5312cfdbb88.png)

Dark (`0.2`):
![image](https://user-images.githubusercontent.com/47470981/146646315-6209a0e0-33be-4d7e-8870-244f1eb13c05.png)

Light (`0.1`):
![image](https://user-images.githubusercontent.com/47470981/146646300-cd18daea-0f1c-43ae-8900-05aa4b6dc172.png)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+            padding: 0.3125rem 0.625rem;
+            flex: 1;
+            background: rgba(0, 0, 0, 0.3);
+            text-align: center;
```
shouldn't be needed, I think.

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+
+        .navbar-default a,
+        >>> .dropdown-toggle {
+            margin: 0 auto;
```
this seems to push the last few item(s) after the dropdown to the right (if there's extra space)

tried removing it and looks good; what's it for?

jonahtanjz replied:
Oh this is for center aligning the text in the default navbar in some screen sizes when the width of each link is slightly larger.

With `margin: 0 auto`:
![image](https://user-images.githubusercontent.com/47470981/145840127-1dae62ac-a0cc-47dd-8d44-55a1a51c2eae.png)

Without `margin: 0 auto`:
![image](https://user-images.githubusercontent.com/47470981/145840049-08c6a240-5e4b-4844-a3e0-9bae359b5118.png)

jonahtanjz replied:
Ahh ok got it. I've flushed all items to the left and it looks alot better.

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+        .navbar-default a,
+        >>> .dropdown-toggle {
+            margin: 0 auto;
+            width: max-content;
```
(just for `>>> .dropdown-toggle`) Could we apply this fix generally to desktop view in dropdown.vue as well? seems to occur there too.

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+        }
+
+        >>> .dropdown-toggle {
+            display: block;
```
per above

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
+
+        >>> .dropdown-toggle {
+            display: block;
+            padding: 0.5rem 0;
```
how about
```
display: flex;
align-items: center;
```
(to avoid tying it to `0.5rem`)

ang-zeyu reviewed (commented):
> I believe to scroll horizontally with the mouse wheel can hold down SHIFT and scroll but this doesn't really seem intuitive 😅 I've bind the mouse wheel to the horizontal scroll on the mobile navbar so this should resolve the issue.

forgot about that 😅, but yes, agree it isn't as intuitive



I only glossed over the positioning / overflow implementation which seems quite complex, it seems to do the job quite well though!

Mostly nits on the styling implementation otherwise:

jonahtanjz commented on packages/vue-components/src/Dropdown.vue:
```diff
@@ -131,11 +132,24 @@ export default {
+      $(this.$refs.dropdown).findChildren('ul').each((ul) => {
+        ul.classList.toggle('show', false);
+
+        if (window.innerWidth < 768 && this.$refs.dropdown.closest('div.navbar-default') !== null) {
```
Yup this seems alot better. Thanks!

jonahtanjz commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +239,98 @@ export default {
     $('.dropdown', this.$el).off('click').offBlur();
-    if (this.slots.collapse) $('[data-toggle="collapse"]', this.$el).off('click');
     $(window).off('resize', this.toggleLowerNavbar);
+    $(this.$el.querySelector('.navbar-default')).off('wheel');
```
Yup using the ref works. I did this part before creating the ref and forgot to update it. Thanks for the catch

jonahtanjz commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -305,6 +361,11 @@ export default {
     /* Navbar link highlight for current page */
+    .navbar .navbar-nav >>> .current:not(.dropdown),
+    .navbar .navbar-nav >>> .dropdown.current {
+        background: none;
```
This is so that when a link in the dropdown is highlighted, the dropdown itself will be highlighted as well.

![image](https://user-images.githubusercontent.com/47470981/145837611-5cf0e207-8db5-4341-9823-097649299ea8.png)

jonahtanjz commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -254,34 +240,97 @@ export default {
+
+        .navbar-right {
+            order: 1;
+            max-width: 50%;
```
> also need to take note 50% does not account for the margin-right set by navbar-brand in bootstrap

Thanks for the catch! I've factored the margin into the `max-width` now and also set `white-space: normal` so that items within `navbar-brand` can wrap to next line when it exceeds `50%`.

> on second thought let's keep in mind and fix separately; found a few more issues that seems to be out of scope here.
But if not we can always keep in mind to to fix soon in the next PR or so, and just focus on the mobile version here.

Yup sure. I can open an issue after this PR to keep track of it :)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -318,8 +332,10 @@ export default {
         .navbar-default > ul > * {
             padding: 0.3125rem 0.625rem;
-            flex: 1;
-            background: rgba(0, 0, 0, 0.3);
```
👀

> ok, I see the point now, I think this works best in that case.

I actually liked the old approach with setting backgroud colour on the individual `ul>*`'s  https://github.com/MarkBind/markbind/pull/1690#discussion_r771130744. (just need a separate `.navbarlight .navbar-default > ul > * { background: ... }` of sorts)

I think dynamic style detection (`--mobile-link-highlight-color`) is something we should avoid when possible as it may introduce more noticeable / distracting visual changes after the page has been presented already. (try seeing the netlify on a slow enough mobile device, you'll get to see the background colour change)

jonahtanjz replied:
Alright I've changed back to the previous implementation. Initially when I flushed all nav items to the left https://github.com/MarkBind/markbind/pull/1690#discussion_r771835529, it left some spaces at the end on some screen sizes so I decided to apply the colour to the entire background instead.

![image](https://user-images.githubusercontent.com/47470981/147409158-77e56ba8-9250-4117-94cd-2047579df54a.png)

I tried playing around with the flex property and decided to go with `flex-grow: 1` so that the nav items fill the bar completely and also ensure that the spacing between each item is "equal". I think this allows for setting the background of each individual nav link as well as addresses the spacing issue.

![image](https://user-images.githubusercontent.com/47470981/147409196-3cefaa2b-1946-4545-982e-5b808aa873bc.png)

ang-zeyu commented on packages/vue-components/src/Navbar.vue:
```diff
@@ -283,17 +288,22 @@ export default {
         div.navbar-brand {
             padding-left: 1rem;
-            max-width: 50%;
+            max-width: calc(50% - 1rem);
```
should we define our own `.navbar-brand` (`.navbar-left`?) class of sorts for use here to avoid relying on `1rem` specifically?

I did a very brief look over the original definition and it seems like it only sets a few paddings / margins / font sizes, nothing too fancy we can't do ourselves 😃

ang-zeyu commented:
one more minor suggestion: should we scroll the scrollbar horizontally toward the "current" element if its hidden beyond the scrollbar?

ang-zeyu commented:
> > Could we preserve the look of the dropdown? (absolutely-positioned with respect to the "Dropdown" text so it dosen't add to the height of the scrollbar) (including the searchbar, which uses some common css classes iirc)
> 
> Yup agree that it will look better. I've updated the design so that it is now absolute positioned and also added scrolling if the dropdown overflows the bottom of the screen.

Missed the search dropdown

jonahtanjz commented:
> one more minor suggestion: should we scroll the scrollbar horizontally toward the "current" element if its hidden beyond the scrollbar?
> 
> It might be somewhat distracting ux wise if it runs too slowly though, you could also try hoisting the scrolling logic into the `<template>` as an inline `<script>` tag so it runs much sooner before the vue app initialisation, if needed.

I tried inserting the scrolling script right after the highlight link function and it seems fine on the preview deployment. The scroll should happen right after the link highlighting with minimal delay.

> Missed the search dropdown

Thanks for the catch! I've also increased the min-width of the search bar on mobile so that it has sufficient space to display the content. Earlier some contents of the search were too compressed.

ang-zeyu reviewed (approved):
Lgtm! 👍
